worker/src/main/java/gov/cms/ab2d/worker/processor/JobDataWriterImpl.java

```diff
@@ -0,0 +1,208 @@
+package gov.cms.ab2d.worker.processor;
+
+import gov.cms.ab2d.common.model.Contract;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static java.nio.file.StandardOpenOption.APPEND;
+
+/**
+ * Since the Worker processes contracts sequentially, create a new instance for each contract,
+ * and make sure only that single instance is shared among the threads.
+ */
+@Slf4j
+public class JobDataWriterImpl implements JobDataWriter {
+    private static final String OUTPUT_FILE_SUFFIX = ".ndjson";
+    private static final String ERROR_FILE_SUFFIX = "_error.ndjson";
+
+    private final Path outputDir;
+    private final Contract contract;
+    private final long maxFileSize;
+    private final int tryLockTimeout;
+
+    private final Lock lock = new ReentrantLock();
+
+    private Path errorFile;
+    private Path dataFile;
+    private final List<Path> dataFiles = new ArrayList<>();
+    private final List<Path> errorFiles = new ArrayList<>();
+
+    private volatile int partitionCounter;
+    private volatile long currentDataFileSize;
+
+
+    public JobDataWriterImpl(Path outputDir, Contract contract, int tryLockTimeout, long maxFileSize) {
+        this.outputDir = outputDir;
+        this.contract = contract;
+        this.tryLockTimeout = tryLockTimeout;
+        this.maxFileSize = maxFileSize;
+    }
+
+
+    /**
+     * Writes a chunk of data to the output file. If needed, transparently rolls the file over to
+     * the next one,
+     * e.g. S00001_0001.ndjson, S00001_0002.ndjson, and so on.
+     * This method MUST be thread-safe and do appropriate locking, e.g. using ReentrantLock
+     *
+     * @param data
+     */
+    @Override
+    public void addDataEntry(byte[] data) {
+        tryLock(lock);
+        try {
+            if (dataFile == null) {
+                createNewDataFile();
+            }
+
+            if (currentDataFileSize + data.length > maxFileSize) {
+                createNewDataFile();
+            }
```

Comment:
What if we simply check the file size on the filesystem? This would eliminate the need to maintain the `currentDataFileSize` variable and remember to increment it when appending, reset to zero when the file is rolled, etc.

Reply:
done